File: backend/schemas/automation.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class CapabilityItem(BaseModel):
    platform: str
    label: str
    oauth: bool
    publish: bool
    publish_media: list[str] = []
    comment_fetch: bool
    comment_reply: bool
    material_variant: bool
    automation_ready: bool
    notes: str = ""
    blockers: list[str] = []


class PolicyUpsert(BaseModel):
    enabled: bool = False
    require_approval: bool = True
    daily_limit: int = Field(default=3, ge=0, le=50)
    auto_reply_enabled: bool = False
    quiet_hours_json: dict[str, Any] | None = None
    config_json: dict[str, Any] | None = None


class PolicyResponse(BaseModel):
    id: uuid.UUID | None = None
    client_id: uuid.UUID
    platform: str
    enabled: bool = False
    require_approval: bool = True
    daily_limit: int = 3
    auto_reply_enabled: bool = False
    quiet_hours_json: dict[str, Any] | None = None
    config_json: dict[str, Any] | None = None
    capability: CapabilityItem | None = None
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


class MaterialGenerateRequest(BaseModel):
    client_id: uuid.UUID
    title: str = Field(..., min_length=1, max_length=500)
    brief: str | None = None
    objective: str | None = None
    target_audience: str | None = None
    core_message: str | None = None
    channels: list[str] = Field(default_factory=lambda: ["instagram", "facebook", "threads"])
    require_approval: bool = True
    generate_images: bool = False
    use_fallback_storyline: bool = False


class ActionLogResponse(BaseModel):
    id: uuid.UUID
    run_id: uuid.UUID | None = None
    client_id: uuid.UUID
    platform: str | None = None
    action: str
    target_id: str | None = None
    ok: bool
    evidence_json: dict[str, Any] | None = None
    error: str | None = None
    created_at: datetime | None = None

    model_config = {"from_attributes": True}


class RunResponse(BaseModel):
    id: uuid.UUID
    client_id: uuid.UUID
    kind: str
    status: str
    summary_json: dict[str, Any] | None = None
    error: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None

    model_config = {"from_attributes": True}


class MaterialGenerateResponse(BaseModel):
    ok: bool
    run: RunResponse
    topic_id: uuid.UUID | None = None
    contents: list[dict[str, Any]] = []
    errors: list[dict[str, Any]] = []


class CommentSyncRequest(BaseModel):
    client_id: uuid.UUID
    platform: str | None = None
    apply_auto_reply: bool = True

File: frontend/src/services/automation.ts
import api from "./api"

export type Capability = {
  platform: string
  label: string
  oauth: boolean
  publish: boolean
  publish_media: string[]
  comment_fetch: boolean
  comment_reply: boolean
  material_variant: boolean
  automation_ready: boolean
  notes: string
  blockers: string[]
}

export type AutomationPolicy = {
  id: string | null
  client_id: string
  platform: string
  enabled: boolean
  require_approval: boolean
  daily_limit: number
  auto_reply_enabled: boolean
  quiet_hours_json?: Record<string, unknown> | null
  config_json?: Record<string, unknown> | null
  capability?: Capability | null
  updated_at?: string | null
}

export type AutomationRun = {
  id: string
  client_id: string
  kind: string
  status: string
  summary_json?: Record<string, unknown> | null
  error?: string | null
  started_at?: string | null
  finished_at?: string | null
}

export type AutomationAction = {
  id: string
  run_id?: string | null
  client_id: string
  platform?: string | null
  action: string
  target_id?: string | null
  ok: boolean
  evidence_json?: Record<string, unknown> | null
  error?: string | null
  created_at?: string | null
}

export type MaterialGenerateResult = {
  ok: boolean
  run: AutomationRun
  topic_id?: string | null
  contents: Array<Record<string, unknown>>
  errors: Array<Record<string, unknown>>
}

export const automationService = {
  async capabilities(): Promise<Capability[]> {
    const res = await api.get("/api/v1/automation/capabilities")
    return Array.isArray(res.data) ? res.data : []
  },

  async listPolicies(clientId: string): Promise<AutomationPolicy[]> {
    const res = await api.get("/api/v1/automation/policies", { params: { client_id: clientId } })
    return Array.isArray(res.data) ? res.data : []
  },

  async upsertPolicy(
    clientId: string,
    platform: string,
    payload: Partial<AutomationPolicy>
  ): Promise<AutomationPolicy> {
    const res = await api.put(`/api/v1/automation/policies/${platform}`, payload, {
      params: { client_id: clientId },
    })
    return res.data
  },

  async generateMaterials(payload: {
    client_id: string
    title: string
    brief?: string
    core_message?: string
    channels: string[]
    require_approval?: boolean
    generate_images?: boolean
    use_fallback_storyline?: boolean
  }): Promise<MaterialGenerateResult> {
    const res = await api.post("/api/v1/automation/materials/generate", payload, { timeout: 190000 })
    return res.data
  },

  async syncComments(clientId: string, platform?: string): Promise<Record<string, unknown>> {
    const res = await api.post(
      "/api/v1/automation/comments/sync",
      { client_id: clientId, platform: platform || null, apply_auto_reply: true },
      { timeout: 120000 }
    )
    return res.data
  },

  async listRuns(clientId: string): Promise<AutomationRun[]> {
    const res = await api.get("/api/v1/automation/runs", { params: { client_id: clientId } })
    return Array.isArray(res.data) ? res.data : []
  },

  async listActions(clientId: string): Promise<AutomationAction[]> {
    const res = await api.get("/api/v1/automation/actions", { params: { client_id: clientId } })
    return Array.isArray(res.data) ? res.data : []
  },
}
